nixpkgs-docs: when to prefer passthru.tests over installCheckPhase (#119731)
NixOS/nixpkgs: doc/stdenv/meta.chapter.md

This attribute is special in that it is not actually under the `meta` attribute set but rather under the `passthru` attribute set. This is due to how `meta` attributes work, and the fact that they are supposed to contain only metadata, not derivations.
:::

An attribute set with tests as values. A test is a derivation that builds when the test passes and fails to build otherwise.

#### Regular tests

> Tests that are part of the source package are often executed in the `installCheckPhase`. Prefer `passthru.tests` for tests that are introduced in nixpkgs because `installCheckPhase` adds overhead to each build and we can run `passthru.tests` independently.

Reviewer: Sorry for reacting a bit late, life is busy ;). I think it would be good to have two motivations here: (1) too much overhead, and (2) where one really wants to test the derivation as a dependency. The problem with `installCheckPhase` is that it still runs within the environment of the derivation, so it may not catch missing dependencies, missing files, etc.

To give an example, suppose that a C library should have another library in its `propagatedBuildInputs`, but it is in the `buildInputs`. You wouldn't detect this with `installCheckPhase`, since it gets the dependency's environment through `buildInputs`. However, a passthru test that builds against the C library would catch the missing propagated build input, since it is not built in the original environment of the derivation.

It is true that downstream users of the C library could also break due to such an error. However, downstream users break for all kinds of reasons, or the maintainer may have added the dependency through `buildInputs`, though it should actually have been propagated.

```nix
{ /* ... */, key }:

mkDerivation rec {
  # ...
  passthru.tests = {
    version-check = runCommand "key-help" {} ''
      ${key}/bin/KeY --help | grep -Fw ${version}
      touch $out
    '';
  };
}
```

> 

Reviewer 2: I [mentioned this out in the main thread](https://github.com/NixOS/nixpkgs/pull/119731#issuecomment-822089022), but I guess I'll attach it here since you pinged me :)

> Without a resource covering what to test, I imagine some readers will see the help/version example here as a straightforward example of a test that is good/desirable or at least acceptable. It might be good to drop some breadcrumbs about where it sits on the various scales (Is it just a trivial example that was once acceptable but would get laughed out of PR review these days? Is it the minimum-meaningful test, or all we should aspire until everything is at this level?)
> 
> If it is trivial, saying so is probably enough. If it is meaningful, a quick high-level note on ~why it's meaningful might be a good stop-gap for orienting people towards appropriate values?

I think it would be good to have a little context on this before or after the example, but I don't see it as a blocker and realize you might not know the answer.

Author: > but I don't see it as a blocker and realize you might not know the answer.

Correct, I don't know if we have enough experience/consensus to be able to put into words authoritative advice on this yet, so I left it out for now.

This attribute is special in that it is not actually under the `meta` attribute set but rather under the `passthru` attribute set. This is due to how `meta` attributes work, and the fact that they are supposed to contain only metadata, not derivations.
:::

> An attribute set with tests as values. A test is a derivation, which builds successfully when the test passes, and fails to build otherwise. A derivation that is a test needs to have `meta.timeout` defined.

Reviewer 3: ```suggestion
An attribute set with tests as values. A test is a derivation, which builds successfully when the test passes, and fails to build otherwise.
```

I think I probably had added the note about `meta.timeout` when initially introducing `passthru.tests`… but it looks like the usage settled on not actually setting it, because it's cumbersome, and it totally makes sense. So maybe let's take advantage of the fixes in this line to remove this sentence, which incidentally doesn't actually match what's done in the paragraph below? :)

#### Regular tests

> Tests that are part of the source package are commonly executed in the `installCheckPhase`. For tests that are introduced in nixpkgs, adding them to `passthru.tests` is preferable because `installCheckPhase` introduces overhead for every build, where `passthru.tests` can be executed independently of the build.

Reviewer 2: ```suggestion
Tests that are part of the source package are often executed in the `installCheckPhase`. Prefer `passthru.tests` for tests added in nixpkgs because `installCheckPhase` adds overhead to each build but we can run `passthru.tests` independently.
```

> Tests that are part of the source package are commonly executed in the `installCheckPhase`. For tests that are introduced in nixpkgs, adding them to `passthru.tests` is preferable because `installCheckPhase` introduces overhead for every build, where `passthru.tests` can be executed independently of the build.

Reviewer 4: @abathur, if you want to cut down on syllables even more:
```suggestion
Tests that are part of the source package are often executed in the `installCheckPhase`. Prefer `passthru.tests` for tests added in nixpkgs. `installCheckPhase` adds overhead to each build, while we can run `passthru.tests` independently.
```

```nix
{ /* ... */ }:
let

>   key = mkDerivation {

Reviewer 2: Does `${version}` require this?

```suggestion
  key = mkDerivation rec {
```

Reviewer 2: I guess this question is too focused on the trees; my forest-thought was whether anything in the example might mislead new packagers.

    # ...
    passthru.tests = {
      version-check = runCommand "key-help" {} ''
        ${key}/bin/KeY --help | grep -Fw ${version}
        touch $out
      '';

>     }
>   }
> };

Reviewer 3: ```suggestion
    };
  };
```

> in key

Reviewer: I am not sure if we should present the example like this, we might end up with a lot of people submitting PRs with ugly `let ... in ...` wrapped derivations, which do not look very canonical. I think it's really much cleaner to just `callPackage` a separate derivation file with tests. Or if it must really be inline pass `pkgs` as an argument (both work because pkgs is a fixed-point).

Also, running a command seems quite lightweight. So, if it is not checked by the regular tests, I think it is easier to just add them to the `postInstallCheck` phase. I think `passthru.tests` is particularly useful for cases where you really want to test use as a separate derivation, e.g. checking whether you can compile against a library and the library works as expected.

Author: > ugly `let ... in ...` wrapped derivations, which do not look very canonical

I agree the `let ... in ...` is rather awkward (I saw it in another passthru).

> I think it's really much cleaner to just `callPackage` a separate derivation file with tests. Or if it must really be inline pass `pkgs` as an argument (both work because pkgs is a fixed-point).

Using `callPackage` (or passing `pkgs` - or perhaps even the package itself?) looks neater, but is a bit dangerous: if you add any overrides to the package, the tests exposed by the override will still test the package without the overrides. If that's a limitation we're happy to live with then that works for me ;).

So perhaps the guidance we should document is: if it's a check that is expected to be quick and can be easily expressed inline, add it to the `installCheckPhase`, otherwise move it to its own file and add it to `passthru.tests`?

Reviewer 3: I'd personally kind of prefer having the `let ... in ...` trick, even though it's not yet canonical, for the reason mentioned by @raboof of actually testing what we think we're testing. As for embedding wrapped derivations inline, if people think it looks bad (which makes sense) one solution would be to have a separate file that looks like:
```nix
# .../tests.nix
{ runCommand` }:
key: version:
{
    version-check = runCommand "key-help" {} ''
      ${key}/bin/KeY --help | grep -Fw ${version}
      touch $out
    '';
}
```

Which would then be called with:
```nix
{ /* ... */ }:
let
  key = mkDerivation {
    # ...
    passthru.tests = callPackage ./tests.nix {} key version;
  }
};
in key
```

Reviewer: I agree that not running tests for overrides is unexpected. But `let self = ...; in self` always felt like accidental complexity that makes derivations much harder to read unless you have in-dept Nix knowledge.

I am wondering if there isn't a more elegant solution. E.g. allowing tests to be `self`-taking functions and having `mkDerivation` pass self?

Reviewer 3: I had that in mind at some point too, but never actually got to it, mostly because anything that touches `mkDerivation` itself I'm trying not to touch — but also because `passthru` is expected to be passed through, so if `mkDerivation` started changing that it would technically be a breaking change.

Overall, I'd be all for such a change! But I don't have the time or energy to see it through myself these days, so… :/

As for the current PR, given that the manual that will be online for the next 6 months is currently being defined (and it's too late to land changes to mkDerivation), I'd think that having this non-elegant workaround in the manual would make sense so people at least know to make tests, and hopefully by the time the next release freezes we will have introduced the necessary mkDerivation machinery?

Reviewer: > As for the current PR, given that the manual that will be online for the next 6 months is currently being defined (and it's too late to land changes to mkDerivation), I'd think that having this non-elegant workaround in the manual would make sense so people at least know to make tests, and hopefully by the time the next release freezes we will have introduced the necessary mkDerivation machinery?

Sounds fair :+1:

Author: > I am wondering if there isn't a more elegant solution. E.g. allowing tests to be `self`-taking functions and having `mkDerivation` pass `self`?

With this in mind I think I'd like to show the approach without `let`, since that'll likely be easier to refactor to the new approach later.

Reviewer 3: FWIW, I'm not sure we'll actually manage to properly do the `self`-taking approach in a reasonable timeframe (I'm pretty sure I personally won't do it anytime soon, and don't know whether other people are planning to); and there may be issues that make it harder than expected (eg. programs that are split between a common.nix and vX-Y.nix, or files that are being callPackage'd multiple times from the top-level with various overrides).

So I wouldn't be too optimistic and would assume that it never actually happens in practice for the documentation we write today, though I'd love to be proven wrong :sweat_smile:

Author: > So I wouldn't be too optimistic and would assume that it never actually happens in practice for the documentation we write today

I agree in general - here it just seemed like a nice tie-breaker ;)

Reviewer 5: This does the trick: https://github.com/NixOS/nixpkgs/pull/119942

> and it's too late to land changes to mkDerivation

I would argue that this is an addition, not a change. It also causes no rebuilds and it's easy to revert.

This attribute is special in that it is not actually under the `meta` attribute set but rather under the `passthru` attribute set. This is due to how `meta` attributes work, and the fact that they are supposed to contain only metadata, not derivations.
:::

An attribute set with tests as values. A test is a derivation that builds when the test passes and fails to build otherwise.

#### Regular tests

Tests that are part of the source package are often executed in the `installCheckPhase`.

Prefer `passthru.tests` for tests that are introduced in nixpkgs because:

* `passthru.tests` tests the 'real' package, independently from the environment in which it was built
* we can run `passthru.tests` independently

> * `installCheckPhase` adds overhead to each build

Reviewer 5: ```suggestion
* `installCheckPhase` adds overhead to each build
* `installCheckPhase` can write to the package
```

https://github.com/NixOS/nixpkgs/issues/120053

Author: (I agree that that is not great, but I don't think it is important enough to list it as a reason to prefer `passthru.tests` here)

```nix

> { /* ... */, key }:

Reviewer 6: ```suggestion
{ ... , key }:
```

Reviewer 5: I disagree. `/* ... */` is clearly intended as meta-syntax, whereas `...` is actual syntax.

Author: @SuperSandro2000 can you clarify why you wanted `/* ... */` to be replaced by `...`, so we can perhaps find a third alternative that is better than the 2 competing ones here?

Author: (I moved this example to doc/contributing/coding-conventions.chapter.md and changed this to `…` (UTF-8 ellipsis) to be consistent with the other already-existing example there)

mkDerivation rec {

>   # ...

Reviewer 6: ```suggestion
  ...
```

Reviewer 5: I disagree. `# ...` is clearly intended as meta-syntax, whereas `...` is close to actual syntax. It's a valid token in the language.

Author: @SuperSandro2000 can you clarify why you wanted `# ...` to be replaced by `...`, so we can perhaps find a third alternative that is better than the 2 competing ones here?

Author: (I moved this example to doc/contributing/coding-conventions.chapter.md and changed this to `…` (UTF-8 ellipsis) to be consistent with the other already-existing example there)

}
```

> NixOS tests are ran in a VM, so they are more expensive than regular tests.

Reviewer 4: ```suggestion
NixOS tests run in a VM, so they are much slower than regular tests.
```
